common/djangoapps/third_party_auth/samlproviderdata/tests/test_samlproviderdata.py

```diff
@@ -0,0 +1,100 @@
+import unittest
+import copy
+import pytz
+from uuid import uuid4
+from datetime import datetime
+from django.contrib.sites.models import Site
+from django.contrib.auth.models import User
+from django.urls import reverse
+from django.utils.http import urlencode
+from rest_framework import status
+from rest_framework.test import APITestCase
+
+from enterprise.models import EnterpriseCustomer, EnterpriseCustomerIdentityProvider
+from enterprise.constants import ENTERPRISE_ADMIN_ROLE
+
+from third_party_auth.tests import testutil
+from third_party_auth.models import SAMLProviderData, SAMLProviderConfig
+from third_party_auth.samlutils.utils import set_jwt_cookie
+
+SINGLE_PROVIDER_CONFIG = {
+    'entity_id': 'http://entity-id-1',
+    'metadata_source': 'http://test.url',
+    'name': 'name-of-config',
+    'enabled': 'true',
+    'slug': 'test-slug'
+}
+
+# entity_id here matches that of the providerconfig, intentionally
+# that allows this data entity to be found
+SINGLE_DATA_CONFIG = {
+    'entity_id': 'http://entity-id-1',
+    'sso_url': 'http://test.url',
+    'public_key': 'a-key0Aid98',
+    'fetched_at': datetime.now(pytz.UTC).replace(microsecond=0)
+}
+
+SINGLE_DATA_CONFIG_2 = copy.copy(SINGLE_DATA_CONFIG)
+SINGLE_DATA_CONFIG_2['entity_id'] = 'http://entity-id-2'
+SINGLE_DATA_CONFIG_2['sso_url'] = 'http://test2.url'
+
+ENTERPRISE_ID = str(uuid4())
+
+
+@unittest.skipUnless(testutil.AUTH_FEATURE_ENABLED, testutil.AUTH_FEATURES_KEY + ' not enabled')
+class SAMLProviderDataTests(APITestCase):
+    """
+        API Tests for SAMLProviderConfig REST endpoints
+    """
+
+    def setUp(self):
+        self.user = User.objects.create_user(username='testuser', password='testpwd')
+        self.site, _ = Site.objects.get_or_create(domain='example.com')
+        self.enterprise_customer = EnterpriseCustomer.objects.create(
+            uuid=ENTERPRISE_ID,
+            name='test-ep',
+            slug='test-ep',
+            site=self.site)
+        self.samlproviderconfig, _ = SAMLProviderConfig.objects.get_or_create(
+            entity_id=SINGLE_PROVIDER_CONFIG['entity_id'],
+            metadata_source=SINGLE_PROVIDER_CONFIG['metadata_source']
+        )
+        # the entity_id here must match that of the samlproviderconfig
+        self.samlproviderdata, _ = SAMLProviderData.objects.get_or_create(
+            entity_id=SINGLE_DATA_CONFIG['entity_id'],
+            sso_url=SINGLE_DATA_CONFIG['sso_url'],
+            fetched_at=SINGLE_DATA_CONFIG['fetched_at']
+        )
+        self.enterprisecustomeridp, _ = EnterpriseCustomerIdentityProvider.objects.get_or_create(
+            provider_id=self.samlproviderconfig.id,
+            enterprise_customer_id=ENTERPRISE_ID
+        )
```

Comment:
You should be able to create all of these in the `setUpTestData` method, so that all of these model instances are created once, not once before each test method.
https://docs.djangoproject.com/en/3.0/topics/testing/tools/#django.test.TestCase.setUpTestData
(`APITestCase` inherits from Django's `TestCase` class: https://www.django-rest-framework.org/api-guide/testing/#api-test-cases)

Reply:
great! I could not figure out how to do the `self.client` part in setUpTestData, since APITestCase seems to be injecting it into the self instance variable only... but moved all other code to setupTestData since it only needs to happen once ...